libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+      return NULL;
+    }
+  
+  return flatpak_remote_ref_action_new (action_type, flatpak_remote_ref, serial);
+}
+
+static gint
+sort_flatpak_remote_ref_actions (gconstpointer a, gconstpointer b)
+{
+  FlatpakRemoteRefAction **action_a = (FlatpakRemoteRefAction **) a;
+  FlatpakRemoteRefAction **action_b = (FlatpakRemoteRefAction **) b;
+
+  return (gint) (*action_a)->serial - (gint) (*action_b)->serial;
+}
+
+static JsonNode *
+parse_json_from_file (GFile         *file,
+                      GCancellable  *cancellable,
+                      GError       **error)
+{
+  g_autoptr(GFileInputStream) input_stream = NULL;
+  g_autoptr(JsonParser) parser = json_parser_new_immutable ();
+
+  input_stream = g_file_read (file, cancellable, error);
+
+  if (!input_stream)
+    return NULL;
+
+  if (!json_parser_load_from_stream (parser,
+                                     G_INPUT_STREAM (input_stream),
+                                     cancellable,
+                                     error))
+    return NULL;
+
+  return json_node_ref (json_parser_get_root (parser));
+}
+
+static const gchar *
+eos_updater_get_system_architecture_string (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_OVERRIDE_ARCHITECTURE",
+                                    flatpak_get_default_arch ());
+}
+
+static gboolean
+architecture_skip_filter_applies (JsonNode  *filter_value,
+                                  gboolean  *is_filtered,
+                                  GError   **error)
+{
+  JsonArray *skip_architectures_array = json_node_get_array (filter_value);
+  g_autoptr(GList) architecture_nodes = NULL;
+  GList *iter = NULL;
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (skip_architectures_array == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (filter_value);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected '~architectures' filter to be an array, was: %s",
+                   node_str);
+      return FALSE;
+    }
+
+  architecture_nodes = json_array_get_elements (skip_architectures_array);
+
+  for (iter = architecture_nodes; iter != NULL; iter = iter->next)
+    {
+      const gchar *string = json_node_get_string (iter->data);
+
+      if (string == NULL)
+        {
+          g_autofree gchar *node_str = json_node_to_string (filter_value);
+          g_set_error (error,
+                       EOS_UPDATER_ERROR,
+                       EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                       "Expected '~architectures' array to contain string values, was: %s",
+                       node_str);
+          return FALSE;
+        }
+
+      if (g_strcmp0 (string, eos_updater_get_system_architecture_string ()) == 0)
+        {
+          *is_filtered = TRUE;
+          return TRUE;
+        }
+    }
+
+  *is_filtered = FALSE;
+  return TRUE;
+}
+
+static gboolean
```

Comment:
All four of these filtering functions are very structurally similar. You could probably merge them into a single function quite easily (which takes a `GStrv` of strings to match against: either the locale list, or a singleton list of the current architecture; and which also takes a `gboolean` telling it whether the match is inverted).

Reply:
I was actually thinking this when I wrote these, just wanted to make sure the tests were passing first before I did this refactor ... then I must not have gotten around to it. I'll do that now, since its relatively trivial.